api/routes/auth.js

```diff
@@ -0,0 +1,120 @@
+const router = require('express').Router()
+const User = require('../models/user')
+const bcrypt = require('bcrypt');
+const saltRounds = 10;
+const jwt = require('jsonwebtoken');
+const { SECRET_KEY } = process.env
+
+router.post('/signup', async (req, res, next) => {
+    const status = 201
+    try {
+        const { username, password } = req.body
+        // Username is not provided
+        if (!username) throw new Error (`Username not provided.`)
+        // Username is already taken
+        const alreadyUsername = await User.findOne({username})
+          .select('-__v -password')
+        if (alreadyUsername) throw new Error (`Username already taken.`)
+
+        // Password is not provided
+        if (!password) throw new Error (`Password not provided.`)
+        // password is less than 8 characters
+        if (password.length < 8) throw new Error (`Password is less than 8 characters.`)
+
+        const hashedPassword = await bcrypt.hash(password, saltRounds)
+        const user = await User.create({
+            username,
+            password: hashedPassword
+        })
+
+        const payload = { id: user._id }
+        const options = { expiresIn: '1 day' }
+        const token = jwt.sign(payload, SECRET_KEY, options)
+
+        res.json({ status, token })    
+    } catch (e) {
+        e.status = 400
+        console.error(e)
+        next(e)
+    }
+})
+
+
+router.post('/login', async (req, res, next) => {
+    const status = 201
+    try {
+        const { username, password } = req.body
+        const user = await User.findOne({ username})
+          .select('-__v -password')
+        // Username is not found
+        if (!user) throw new Error (`Incorrect login credentials.`)
+        
+        // If username does exist, compare the plain text password the the hashed version
+        const isValid = await bcrypt.compare(password, user.password)
+        if (!isValid) throw new Error(`Incorrect login credentials.`)
+
+        const payload = { id: user._id }
+        const options = { expiresIn: '1 day' }
+        const token = jwt.sign(payload, SECRET_KEY, options)
```

Comment:
Because you're performing this operation multiple times, this is a great thing to put into a function.